cmd/npu_plugin/npu_plugin.go

```diff
@@ -0,0 +1,391 @@
+	policy  preferredAllocationPolicyFunc
+	options cliOptions
+}
+
+func newDevicePlugin(sysfsDir, devfsDir string, options cliOptions) *devicePlugin {
+	dp := &devicePlugin{
+		sysfsDir:     sysfsDir,
+		devfsDir:     devfsDir,
+		options:      options,
+		npuDeviceReg: regexp.MustCompile(npuDeviceRE),
+		scanTicker:   time.NewTicker(scanPeriod),
+		scanDone:     make(chan bool, 1), // buffered as we may send to it before Scan starts receiving from it
+	}
+
+	switch options.preferredAllocationPolicy {
+	case "balanced":
+		dp.policy = balancedPolicy
+	case "packed":
+		dp.policy = packedPolicy
+	default:
+		dp.policy = nonePolicy
+	}
+
+	return dp
+}
+
+// Implement the PreferredAllocator interface.
+func (dp *devicePlugin) GetPreferredAllocation(rqt *pluginapi.PreferredAllocationRequest) (*pluginapi.PreferredAllocationResponse, error) {
+	response := &pluginapi.PreferredAllocationResponse{}
+
+	for _, req := range rqt.ContainerRequests {
+		klog.V(3).Infof("AvailableDeviceIDs: %q", req.AvailableDeviceIDs)
+		klog.V(3).Infof("MustIncludeDeviceIDs: %q", req.MustIncludeDeviceIDs)
+		klog.V(3).Infof("AllocationSize: %d", req.AllocationSize)
+
+		// Add a security check here. This should never happen unless there occurs error in kubelet device plugin manager.
+		if req.AllocationSize > int32(len(req.AvailableDeviceIDs)) {
+			klog.V(3).Info("req.AllocationSize must be not greater than len(req.AvailableDeviceIDs).")
+
+			var err = errors.Errorf("AllocationSize (%d) is greater than the number of available device IDs (%d)", req.AllocationSize, len(req.AvailableDeviceIDs))
+
+			return nil, err
+		}
+
+		IDs := dp.policy(req)
+
+		resp := &pluginapi.ContainerPreferredAllocationResponse{
+			DeviceIDs: IDs,
+		}
+
+		response.ContainerResponses = append(response.ContainerResponses, resp)
+	}
+
+	return response, nil
+}
+
+func (dp *devicePlugin) Scan(notifier dpapi.Notifier) error {
+	defer dp.scanTicker.Stop()
+
+	klog.V(1).Infof("NPU (%s) resource share count = %d", deviceTypeNpu, dp.options.sharedDevNum)
+
+	previousCount := 0
+	devType := fmt.Sprintf("%s/%s", namespace, deviceTypeNpu)
+
+	for {
+		devTree, err := dp.scan()
+		if err != nil {
+			klog.Errorf("NPU scan failed: %v", err)
+			return errors.Wrap(err, "NPU scan failed")
+		}
+
+		count := devTree.DeviceTypeCount(devType)
+		if count != previousCount {
+			klog.V(1).Infof("NPU scan update: %d->%d '%s' resources found", previousCount, count, devType)
+
+			previousCount = count
+		}
+
+		notifier.Notify(devTree)
+
+		select {
+		case <-dp.scanDone:
+			return nil
+		case <-dp.scanTicker.C:
+		}
+	}
+}
+
+func (dp *devicePlugin) isCompatibleDevice(name string) bool {
+	if !dp.npuDeviceReg.MatchString(name) {
+		klog.V(4).Info("Not compatible device: ", name)
+		return false
+	}
+
+	dat, err := os.ReadFile(path.Join(dp.sysfsDir, name, "device/vendor"))
+	if err != nil {
+		klog.Warning("Skipping. Can't read vendor file: ", err)
```

Comment:
I wonder whether this needs to be a warning. E.g. KMD for Gaudi accelerators does not provide `vendor` file (nor `device` file), and there could be other accelerator devices that do not provide it either...

(And if they would provide device file,